### PR TITLE
Scaffold RoslynSymbolicExecution

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Only commits after `Squash SimplifyRunner` needs a review

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs [hunk 1]
```diff
@@ -32,7 +33,14 @@ public class RoslynSymbolicExecution
 
         public void Execute()
         {
-            // ToDo: Something is missing here
+            foreach (var block in cfg.Blocks)    // ToDo: This is a temporary simplification until we support proper branching
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs]: great!

I love simplifications 😄

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs [hunk 2]
```diff
@@ -27,20 +28,47 @@ namespace SonarAnalyzer.SymbolicExecution.Roslyn
...
         public void Execute()
         {
-            foreach (var block in cfg.Blocks)    // ToDo: This is a temporary simplification until we support proper branching
+            foreach (var block in cfg.Blocks)   // ToDo: This is a temporary simplification until we support proper branching
             {
-                foreach(var operation in block.OperationsAndBranchValue.ToExecutionOrder()) // ToDo: This is a temporary oversimplification to scaffold the engine
+                var state = new ProgramState(); // ToDo: This is a temporary oversimplification to scaffold the enginge
+                foreach (var operation in block.OperationsAndBranchValue.ToExecutionOrder()) // ToDo: This is a temporary oversimplification to scaffold the engine
                 {
+                    state = InvokeChecks(state, (check, ps) => check.PreProcess(ps, operation));
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:50]: silly question: shouldn't this return a list of states?

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:50]: _In theory:_ yes.
_In practice:_ no, it's not needed by our rules.
_In the future:_ we can always upgrade if we need to.

File: analyzers/src/SonarAnalyzer.Common/SonarAnalyzer.Common.csproj [hunk 3]
```diff
@@ -8,6 +8,12 @@
     <ProjectGuid>{8a8a663e-1318-4361-bc97-2e63666feadb}</ProjectGuid>
   </PropertyGroup>
 
+  <ItemGroup>
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SonarAnalyzer.Common.csproj:11]: what is this about?

should it be removed?

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicExecutionCheck.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using StyleCop.Analyzers.Lightup;
+
+namespace SonarAnalyzer.SymbolicExecution.Roslyn
+{
+    public class SymbolicExecutionCheck
+    {
+        protected SymbolicExecutionCheck() { } // Avoid abstract class
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicExecutionCheck.cs:27]: >Avoid abstract class

what do you mean here?

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM

(please check `SonarAnalyzer.Common.csproj` before merge)

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs]: 😅

File: analyzers/src/SonarAnalyzer.Common/SonarAnalyzer.Common.csproj (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SonarAnalyzer.Common.csproj:11]: Autogenerated noise

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicExecutionCheck.cs (hunk 4 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicExecutionCheck.cs:27]: It fixes S1694, I'll add a comment

**Pull request merged**